src/lookup.h

```diff
@@ -0,0 +1,267 @@
+            }
+        }
+    }
+
+    size_t usedMemory() {
+        size_t bytes = 0;
+        for (const_iterator it = begin(); it != end(); ++it) {
+            bytes += sizeof(jmethodID) + sizeof(MethodInfo);
+            bytes += it->second._line_number_table_size * sizeof(jvmtiLineNumberEntry);
+        }
+        return bytes;
+    }
+};
+
+class Lookup {
+  public:
+    MethodMap* _method_map;
+    // Dictionary is thread-safe and can be safely shared, as opposed to Index
+    Dictionary* _classes;
+    Index* _packages;
+    Index* _symbols;
+
+  private:
+    JNIEnv* _jni;
+    bool _owns_packages = false;
+    bool _owns_symbols = false;
+
+    void fillNativeMethodInfo(MethodInfo* mi, const char* name, const char* lib_name) {
+        if (lib_name == NULL) {
+            mi->_file = _classes->lookup("");
+        } else if (lib_name[0] == '[' && lib_name[1] != 0) {
+            mi->_file = _classes->lookup(lib_name + 1, strlen(lib_name) - 2);
+        } else {
+            mi->_file = _classes->lookup(lib_name);
+        }
+
+        mi->_modifiers = 0x100;
+        mi->_line_number_table_size = 0;
+        mi->_line_number_table = NULL;
+
+        mi->_system_name = _symbols->indexOf(name);
+        if (Demangle::needsDemangling(name)) {
+            char* demangled = Demangle::demangle(name, false);
+            if (demangled != NULL) {
+                mi->_name = _symbols->indexOf(demangled);
+                mi->_sig = _symbols->indexOf("()L;");
+                mi->_type = FRAME_CPP;
+                free(demangled);
+                return;
+            }
+        }
+
+        size_t len = strlen(name);
+        if (len >= 4 && strcmp(name + len - 4, "_[k]") == 0) {
+            mi->_name = _symbols->indexOf(name, len - 4);
+            mi->_sig = _symbols->indexOf("(Lk;)L;");
+            mi->_type = FRAME_KERNEL;
+        } else {
+            mi->_name = _symbols->indexOf(name);
+            mi->_sig = _symbols->indexOf("()L;");
+            mi->_type = FRAME_NATIVE;
+        }
+    }
+
+    bool fillJavaMethodInfo(MethodInfo* mi, jmethodID method, bool first_time) {
+        if (VMMethod::isStaleMethodId(method)) {
+            return false;
+        }
+
+        jclass method_class = NULL;
+        char* class_name = NULL;
+        char* method_name = NULL;
+        char* method_sig = NULL;
+        char* source_name = NULL;
+
+        jvmtiEnv* jvmti = VM::jvmti();
+        jvmtiError err;
+
+        if ((err = jvmti->GetMethodName(method, &method_name, &method_sig, NULL)) == 0 &&
+            (err = jvmti->GetMethodDeclaringClass(method, &method_class)) == 0 &&
+            (err = jvmti->GetClassSignature(method_class, &class_name, NULL)) == 0 &&
+            (err = jvmti->GetSourceFileName(method_class, &source_name)) == 0) {
+            mi->_class = _classes->lookup(class_name + 1, strlen(class_name) - 2);
+            mi->_file = _classes->lookup(source_name);
```

Comment:
This adds an entry in the classes dictionary. This name will then appear as a class in JMC or other JFR viewer.